File: unila-aqi-backend/src/models/SensorData.js
const mongoose = require('mongoose');

const sensorDataSchema = new mongoose.Schema({
  roomId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Room',
    required: true
  },
  roomName: {
    type: String,
    required: true
  },
  buildingName: {
    type: String,
    required: true
  },
  aqi: {
    type: Number,
    required: true
  },
  pm25: {
    type: Number,
    required: true
  },
  pm10: {
    type: Number,
    required: true
  },
  co2: {
    type: Number,
    required: true
  },
  temperature: {
    type: Number,
    required: true
  },
  humidity: {
    type: Number,
    required: true
  },
  category: {
    type: String,
    enum: ['baik', 'sedang', 'tidak_sehat', 'sangat_tidak_sehat', 'berbahaya', 'error'], // TAMBAH 'error'
    required: true
  },
  timestamp: {
    type: Date,
    default: Date.now,
    index: true
  }
});

// Create index untuk query yang cepat
sensorDataSchema.index({ roomId: 1, timestamp: -1 });
sensorDataSchema.index({ timestamp: -1 });

module.exports = mongoose.model('SensorData', sensorDataSchema);

File: unila_aqi/lib/data/models/sensor_data.dart
class SensorData {
  final String id;
  final String roomId;
  final String roomName;
  final String buildingName;
  final int aqi;
  final double pm25;
  final double pm10;
  final double co2;
  final double temperature;
  final double humidity;
  final String category;
  final DateTime timestamp;
  final List<dynamic>? rawData; // Untuk data agregasi
  final String? timeLabel; // Untuk label waktu

  SensorData({
    required this.id,
    required this.roomId,
    required this.roomName,
    required this.buildingName,
    required this.aqi,
    required this.pm25,
    required this.pm10,
    required this.co2,
    required this.temperature,
    required this.humidity,
    required this.category,
    required this.timestamp,
    this.rawData,
    this.timeLabel,
  });

  factory SensorData.fromJson(Map<String, dynamic> json) {
    return SensorData(
      id: json['_id']?.toString() ?? json['id']?.toString() ?? DateTime.now().millisecondsSinceEpoch.toString(),
      roomId: json['roomId']?.toString() ?? '',
      roomName: json['roomName']?.toString() ?? '',
      buildingName: json['buildingName']?.toString() ?? '',
      aqi: (json['aqi'] ?? 0).toInt(),
      pm25: (json['pm25'] ?? 0).toDouble(),
      pm10: (json['pm10'] ?? 0).toDouble(),
      co2: (json['co2'] ?? 0).toDouble(),
      temperature: (json['temperature'] ?? 0).toDouble(),
      humidity: (json['humidity'] ?? 0).toDouble(),
      category: json['category']?.toString() ?? 'baik',
      timestamp: json['timestamp'] != null 
          ? DateTime.parse(json['timestamp'].toString()).toLocal()
          : DateTime.now(),
      rawData: json['rawData'],
      timeLabel: json['timeLabel']?.toString(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'roomId': roomId,
      'roomName': roomName,
      'buildingName': buildingName,
      'aqi': aqi,
      'pm25': pm25,
      'pm10': pm10,
      'co2': co2,
      'temperature': temperature,
      'humidity': humidity,
      'category': category,
      'timestamp': timestamp.toIso8601String(),
      'rawData': rawData,
      'timeLabel': timeLabel,
    };
  }
}